.github/workflows/benchmark.yml

```
@@ -1,50 +1,267 @@
-      - uses: actions/checkout@v4
+      - name: Check out code
+        uses: actions/checkout@v4
         with:
           fetch-depth: 0
-      - name: jemalloc dependency
-        run: apt-get update && apt-get install -y libjemalloc-dev
-      - name: Fix Git config
-        run: |
-          git config --global --add safe.directory "${GITHUB_WORKSPACE}"
-      - name: Run benchmarks for PR branch
-        continue-on-error: true
-        run: |
-          swift package -c release --package-path Benchmarks --disable-sandbox benchmark baseline update pull_request --no-progress --quiet
-      - name: Run benchmarks for 'main' branch
-        run: |
-          git stash
-          git checkout main
-          swift package -c release --package-path Benchmarks --disable-sandbox benchmark baseline update main --no-progress --quiet
-      - name: Compare benchmarks
-        continue-on-error: true
-        run: |
-          date >> "${GITHUB_STEP_SUMMARY}"
-          swift package -c release --package-path Benchmarks benchmark baseline check main pull_request --format markdown >> "${GITHUB_STEP_SUMMARY}"
-      - name: Get formatted date
-        id: get-date
-        run: echo "date=$(date +'%Y-%m-%d %H:%M:%S')" >> $GITHUB_OUTPUT
-      - uses: thollander/actions-comment-pull-request@v3
+
+      - name: Configure git
+        run: git config --global --add safe.directory "${GITHUB_WORKSPACE}"
+
+      # jemalloc is a dependency of the Benchmarking package
+      # actions/cache will detect zstd and will become much faster.
+      - name: Install jemalloc, curl, jq and zstd
+        run: |
+          set -eu
+
+          apt-get update -y
+          apt-get install -y libjemalloc-dev curl jq zstd
+
+      - name: Restore .build
+        id: "restore-cache"
+        uses: runs-on/cache/restore@v4
         with:
-          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
-          message: ${{ format('[PR benchmark comparison with main on ubuntu-latest run at {0}]({1}/{2}/actions/runs/{3})', steps.get-date.outputs.date, github.server_url, github.repository, github.run_id) }}
-          comment_tag: 'PR benchmark comparison Linux'
+          path: Benchmarks/.build
+          key: "swiftpm-benchmark-build-${{ runner.os }}-${{ github.event.pull_request.base.sha || github.event.after }}"
+          restore-keys: "swiftpm-benchmark-build-${{ runner.os }}-"
+
+      - name: Run Benchmarks Of Current Branch '${{ github.head_ref || github.ref_name }}'
+        run: |
+          swift package -c release --disable-sandbox \
+            --package-path Benchmarks \
+            benchmark baseline update \
+            "${{ github.head_ref || github.ref_name }}"
+
+      - name: Read Benchmark Result
+        id: read-benchmark
+        run: |
+          set -eu
+
+          swift package -c release --disable-sandbox \
+            --package-path Benchmarks \
+            benchmark baseline read \
+            "${{ github.head_ref || github.ref_name }}" \
```

Don't allow interpolation of workflow-inserted text.
```suggestion
            '${{ github.head_ref || github.ref_name }}' \
```